perfmetrics/scripts/run_e2e_tests.sh

```diff
@@ -138,69 +148,146 @@ function run_parallel_tests() {
+  echo "Running parallel tests..."
+  # Run parallel tests
+  run_parallel_tests TEST_DIR_PARALLEL $BUCKET_NAME_PARALLEL &
+  parallel_tests_pid=$!
+
+ # Non parallel execution of integration tests located within specified test directories.
+ echo "Running non parallel tests group-1..."
+ run_non_parallel_tests TEST_DIR_NON_PARALLEL_GROUP_1 $BUCKET_NAME_NON_PARALLEL_GROUP_1 &
+ non_parallel_tests_pid_group_1=$!
+ echo "Running non parallel tests group-2..."
+ run_non_parallel_tests TEST_DIR_NON_PARALLEL_GROUP_2 $BUCKET_NAME_NON_PARALLEL_GROUP_2 &
+ non_parallel_tests_pid_group_2=$!
+
+ # Wait for all tests to complete.
+ wait $parallel_tests_pid
+ parallel_tests_exit_code=$?
+ wait $non_parallel_tests_pid_group_1
+ non_parallel_tests_exit_code_group_1=$?
+ wait $non_parallel_tests_pid_group_2
+ non_parallel_tests_exit_code_group_2=$?
+
+
+ if [ $non_parallel_tests_exit_code_group_1 != 0 ] || [ $non_parallel_tests_exit_code_group_2 != 0 ] || [ $parallel_tests_exit_code != 0 ];
+ then
+   exit 1
+ fi
+}
+
+function run_e2e_tests_for_hns_bucket(){
+   echo "Running tests for HNS bucket"
+   run_non_parallel_tests TEST_DIR_HNS_GROUP "$HNS_BUCKET_NAME"
+   non_parallel_tests_pid_hns_group=$!
+
+   wait $non_parallel_tests_pid_hns_group
+   non_parallel_tests_hns_group_exit_code=$?
+
+   if [ $non_parallel_tests_hns_group_exit_code != 0 ];
+   then
+     exit 1
+   fi
+}
+
+#commenting it so cleanup and failure check happens for both
+#set -e
+
+function clean_up() {
+  # Cleanup
+  # Delete bucket after testing.
+  local -n buckets=$1
+  for bucket in "${buckets[@]}"
+    do
+      gcloud alpha storage rm --recursive gs://$bucket
+    done
+}
+
+function clean_up_buckets(){
+  flat_buckets=("$BUCKET_NAME_PARALLEL" "$BUCKET_NAME_NON_PARALLEL_GROUP_1" "$BUCKET_NAME_NON_PARALLEL_GROUP_2")
+  hns_buckets=("$HNS_BUCKET_NAME")
+
+  clean_up flat_buckets
+  clean_up hns_buckets
+}
+
+function run_tests() {
+  run_e2e_tests_for_flat_bucket &
+  e2e_tests_flat_bucket_pid=$!
+
+  run_e2e_tests_for_hns_bucket &
+  e2e_tests_hns_bucket_pid=$!
+
+  wait $e2e_tests_flat_bucket_pid
+  e2e_tests_flat_bucket_status=$?
+
+  wait $e2e_tests_hns_bucket_pid
+  e2e_tests_hns_bucket_status=$?
+
+  if [ $e2e_tests_flat_bucket_status != 0 ];
+  then
+    echo "The e2e tests for flat bucket failed.."
+  fi
+
+  if [ $e2e_tests_hns_bucket_status != 0 ];
+  then
+    echo "The e2e tests for hns bucket failed.."
+  fi
+}
+
+function main(){
+  set -e
+
+  upgrade_gcloud_version
+
+  install_packages
+
+  create_flat_buckets
+  HNS_BUCKET_NAME=$(create_hns_bucket)
+  echo "Hns Bucket Created: "$HNS_BUCKET_NAME
+
+  set +e
+
+  #run integration tests
+  run_tests
+
+  # Cleanup
+  # Delete bucket after testing.
+  clean_up_buckets
+
+  # Removing bin file after testing.
+  if [ $RUN_E2E_TESTS_ON_PACKAGE != true ];
+  then
+    sudo rm /usr/local/bin/gcsfuse
+  fi
+}
+
+#Main method to run script
+main
```

Comment:
question - does the script fails if something fails inside main, asking because we are not returning any exit code from main, also "set +e" is set inside main before run_tests is called.

Reply:
its a bug !! script will not fail if any statement errors out